server/odc-service/src/main/java/com/oceanbase/odc/service/sqlcheck/rule/MySQLAffectedRows.java

```diff
@@ -0,0 +1,182 @@
+
+    public MySQLAffectedRows(@NonNull Integer maxSQLAffectedRows, JdbcOperations jdbcOperations) {
+        this.maxSQLAffectedRows = maxSQLAffectedRows <= 0 ? 0 : maxSQLAffectedRows;
+        this.jdbcOperations = jdbcOperations;
+    }
+
+    /**
+     * Get the rule type
+     */
+    @Override
+    public SqlCheckRuleType getType() {
+        return SqlCheckRuleType.RESTRICT_SQL_AFFECTED_ROWS;
+    }
+
+    /**
+     * Execution rule check
+     */
+    @Override
+    public List<CheckViolation> check(@NonNull Statement statement, @NonNull SqlCheckContext context) {
+
+        if (statement instanceof Insert || statement instanceof Update || statement instanceof Delete) {
+            if (maxSQLAffectedRows == 0) {
+                return Collections.emptyList();
+            }
+            String originSql = statement.getText();
+            long affectedRows = getAffectedRows(originSql, jdbcOperations);
+            if (affectedRows > maxSQLAffectedRows) {
+                return Collections.singletonList(SqlCheckUtil
+                        .buildViolation(statement.getText(), statement, getType(),
+                                new Object[] {maxSQLAffectedRows, affectedRows}));
+            } else {
+                return Collections.emptyList();
+            }
+        }
+        return Collections.emptyList();
+    }
+
+    /**
+     * Get supported database types
+     */
+    @Override
+    public List<DialectType> getSupportsDialectTypes() {
+        return Arrays.asList(DialectType.MYSQL, DialectType.OB_MYSQL);
+    }
+
+    /**
+     * execute 'explain' statement
+     *
+     * @param originSql target sql
+     * @param jdbc jdbc Object
+     * @return affected rows
+     */
+    private long getAffectedRows(String originSql, JdbcOperations jdbc) {
+
+        if (originSql == null || jdbc == null) {
+            return 0;
+        }
+        String explainSql = "EXPLAIN " + originSql;
+
+        /**
+         * <pre>
+         *
+         *     explain result (json):
+         *
+         *     ==================================================    --rowNum 1
+         *     |ID|OPERATOR          |NAME|EST.ROWS|EST.TIME(us)|    --rowNum 2
+         *     0 |DISTRIBUTED UPDATE|    |1       |37          |     --rowNum 3
+         *     1 |└─TABLE GET       |user|1       |5           |     --rowNum 4
+         *     ==================================================    --rowNum 5
+         *     ...
+         *
+         * </pre>
+         */
+        try {
+            AtomicBoolean ifFindAffectedRow = new AtomicBoolean(false);
+            List<Long> resultSet = jdbc.query(explainSql, (rs, rowNum) -> {
+
+                String resultRow = rs.getString("Query Plan");
+                if (!ifFindAffectedRow.get() && rowNum > HEAD_LINE) {
+                    long affectedRows = getEstRowsValue(resultRow);
+                    // first non-null value is the column 'EST.ROWS'
+                    if (affectedRows != 0) {
```

Comment:
return 0 seems not a good design. What if the 'affected rows' is indeed 0?

Reply:
Yes, I realized this problem, and the latesr code has been modified.